swim/heal_via_discover_provider.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import (
+	"math/rand"
+	"time"
+
+	log "github.com/uber-common/bark"
+	"github.com/uber/ringpop-go/logging"
+)
+
+// discoverProviderHealer attempts to heal a ringpop partition by consulting
+// the discoverProvider for nodes that are faulty or not available in its
+// membership. It attempts the heal probabalisticly so that the discovery provider
+// doesn't get overloaded with request -- with default settings 6
+// times per minutes on avarage for the entire cluster.
+type discoverProviderHealer struct {
+	node *Node
+
+	baseProbabillity float64
+	period           time.Duration
+
+	previousHostListSize int
+	quit                 chan struct{}
+
+	logger log.Logger
+}
+
+func newDiscoverProviderHealer(n *Node, baseProbability float64, period time.Duration) *discoverProviderHealer {
+	return &discoverProviderHealer{
+		node:             n,
+		baseProbabillity: baseProbability,
+		period:           period,
+		quit:             make(chan struct{}),
```

Comment:
With the `quit` channel initialized here only once there is no way to `Start` the healer again after it `Stop` has been called on it. This would be the first submodule of `Node` that has this behavior, both the `gossip` and `stateTransitition` modules can be stopped and started multiple times through their life time.